File: showcase/integrations/claude-sdk-python/src/agents/claude_agent_sdk_adapter.py
"""Official Claude Agent SDK adapter wiring for the showcase agents."""

from __future__ import annotations

import os
from collections.abc import AsyncIterator, Callable
from typing import Any

from ag_ui.core import EventType, RunAgentInput, StateSnapshotEvent
from ag_ui.encoder import EventEncoder
from ag_ui_claude_sdk import ClaudeAgentAdapter
from claude_agent_sdk import create_sdk_mcp_server, tool as sdk_tool

COPILOTKIT_MCP_SERVER_NAME = "copilotkit"
COPILOTKIT_TOOL_PREFIX = f"mcp__{COPILOTKIT_MCP_SERVER_NAME}__"

ExecuteTool = Callable[
    [str, dict[str, Any], Any, list[dict[str, Any]] | None],
    tuple[str, Any | None],
]


def should_use_claude_agent_sdk(
    *,
    input_data: RunAgentInput,
    backend_tools: list[dict[str, Any]],
    frontend_tool_names: set[str],
    preprocess_user_parts: Any = None,
) -> bool:
    """Return whether this request can safely run through the official adapter."""

    base_url = os.getenv("ANTHROPIC_BASE_URL", "")
    if "aimock" in base_url:
        return False

    # The official adapter uses the Claude Agent SDK process transport, which
    # has no supported per-request HTTP header hook for x-aimock-context today.
    if preprocess_user_parts is not None:
        return False

    # The official Claude Agent SDK path can execute backend MCP tools, but it
    # does not yet bridge CopilotKit frontend/runtime tools back through AG-UI.
    if frontend_tool_names:
        return False

    if _has_structured_user_content(input_data):
        return False

    return True


# @region[claude-agent-sdk-python-adapter]
async def run_with_claude_agent_sdk(
    input_data: RunAgentInput,
    *,
    system_prompt: str,
    tools: list[dict[str, Any]],
    state: Any,
    model: str,
    execute_tool: ExecuteTool,
    max_turns: int = 10,
) -> AsyncIterator[str]:
    """Run through the official AG-UI Claude adapter and emit SSE chunks."""

    # @region[claude-agent-sdk-agent-setup]
    encoder = EventEncoder()
    state_box = {"state": state}
    pending_state_snapshots: list[Any] = []
    sdk_tools = _build_sdk_tools(
        tools,
        execute_tool=execute_tool,
        get_state=lambda: state_box["state"],
        set_state=lambda next_state: _set_state(
            next_state,
            state_box,
            pending_state_snapshots,
        ),
    )

    options: dict[str, Any] = {
        "model": _normalize_claude_agent_sdk_model(model),
        "system_prompt": system_prompt,
        "tools": [],
        "permission_mode": "dontAsk",
        "max_turns": max_turns,
    }

    if sdk_tools:
        options["mcp_servers"] = {
            COPILOTKIT_MCP_SERVER_NAME: create_sdk_mcp_server(
                COPILOTKIT_MCP_SERVER_NAME,
                "1.0.0",
                tools=sdk_tools,
            )
        }
        options["allowed_tools"] = [
            f"{COPILOTKIT_TOOL_PREFIX}{schema['name']}" for schema in tools
        ]

    adapter = ClaudeAgentAdapter(
        name="claude-sdk-python",
        options=options,
    )
    run_input = _with_initial_state(input_data, state)

    async for event in adapter.run(run_input):
        if event.type == EventType.TOOL_CALL_RESULT and pending_state_snapshots:
            yield encoder.encode(
                StateSnapshotEvent(
                    type=EventType.STATE_SNAPSHOT,
                    snapshot=pending_state_snapshots.pop(0),
                )
            )
        yield encoder.encode(event)
    # @endregion[claude-agent-sdk-agent-setup]


def _build_sdk_tools(
    tool_schemas: list[dict[str, Any]],
    *,
    execute_tool: ExecuteTool,
    get_state: Callable[[], Any],
    set_state: Callable[[Any], None],
) -> list[Any]:
    return [
        _make_sdk_tool(
            schema,
            execute_tool=execute_tool,
            get_state=get_state,
            set_state=set_state,
        )
        for schema in tool_schemas
    ]


def _make_sdk_tool(
    schema: dict[str, Any],
    *,
    execute_tool: ExecuteTool,
    get_state: Callable[[], Any],
    set_state: Callable[[Any], None],
) -> Any:
    name = schema["name"]
    description = schema.get("description", "")
    input_schema = schema.get("input_schema", {"type": "object", "properties": {}})

    @sdk_tool(name, description, input_schema)
    async def sdk_tool_handler(args: dict[str, Any]):
        try:
            result_text, next_state = execute_tool(
                name,
                dict(args or {}),
                get_state(),
                None,
            )
            if next_state is not None:
                set_state(next_state)
            return {"content": [{"type": "text", "text": result_text}]}
        except Exception as exc:
            return {
                "content": [{"type": "text", "text": str(exc)}],
                "is_error": True,
            }

    return sdk_tool_handler


def _set_state(
    next_state: Any,
    state_box: dict[str, Any],
    pending_state_snapshots: list[Any],
) -> None:
    state_box["state"] = next_state
    pending_state_snapshots.append(_snapshot_from_state(next_state))


def _snapshot_from_state(state: Any) -> Any:
    if hasattr(state, "model_dump"):
        return state.model_dump()
    return state


def _with_initial_state(input_data: RunAgentInput, state: Any) -> RunAgentInput:
    if getattr(input_data, "state", None) is not None or state is None:
        return input_data
    if hasattr(input_data, "model_copy"):
        return input_data.model_copy(update={"state": state})
    if hasattr(input_data, "copy"):
        return input_data.copy(update={"state": state})
    return input_data


def _has_structured_user_content(input_data: RunAgentInput) -> bool:
    for msg in input_data.messages or []:
        role = msg.role.value if hasattr(msg.role, "value") else str(msg.role)
        raw_content = getattr(msg, "content", None)
        if role == "user" and isinstance(raw_content, list):
            for part in raw_content:
                part_type = getattr(part, "type", None) or (
                    part.get("type") if isinstance(part, dict) else None
                )
                if part_type and part_type != "text":
                    return True
    return False


def _normalize_claude_agent_sdk_model(model: str) -> str:
    return "claude-sonnet-4-6" if model == "claude-sonnet-4.6" else model


# @endregion[claude-agent-sdk-python-adapter]

File: showcase/integrations/claude-sdk-typescript/src/claude-agent-sdk-adapter.ts
import { ClaudeAgentAdapter } from "@ag-ui/claude-agent-sdk";
import { EventType } from "@ag-ui/core";
import type { BaseEvent, RunAgentInput } from "@ag-ui/core";
import {
  createSdkMcpServer,
  tool as sdkTool,
} from "@anthropic-ai/claude-agent-sdk";
import type {
  AnyZodRawShape,
  McpServerConfig,
} from "@anthropic-ai/claude-agent-sdk";
import type Anthropic from "@anthropic-ai/sdk";
import { z } from "zod4/v4";
import type { ZodTypeAny } from "zod4/v4";

const COPILOTKIT_MCP_SERVER_NAME = "copilotkit";
const COPILOTKIT_TOOL_PREFIX = `mcp__${COPILOTKIT_MCP_SERVER_NAME}__`;

type Emit = (event: BaseEvent | object) => void;

type ExecuteToolResult = {
  resultText: string;
  state: Record<string, unknown> | null;
};

type ExecuteTool = (
  toolName: string,
  toolInput: Record<string, unknown>,
  state: Record<string, unknown>,
  emit: Emit,
) => Promise<ExecuteToolResult>;

export function shouldUseClaudeAgentSdk({
  input,
  forwardedHeaders,
  runtimeToolCount,
  enableThinking,
}: {
  input: RunAgentInput;
  forwardedHeaders: Record<string, string>;
  runtimeToolCount: number;
  enableThinking?: boolean;
}): boolean {
  if ((process.env.ANTHROPIC_BASE_URL ?? "").includes("aimock")) {
    return false;
  }
  // The official adapter keeps a `headers` property for forward compatibility,
  // but the Claude Agent SDK cannot forward per-request HTTP headers today.
  if (hasHeader(forwardedHeaders, "x-aimock-context")) {
    return false;
  }
  if (enableThinking) {
    return false;
  }
  // The official Claude Agent SDK path can execute backend MCP tools, but it
  // does not yet bridge CopilotKit frontend/runtime tools back through AG-UI.
  if (runtimeToolCount > 0) {
    return false;
  }
  if (hasStructuredUserContent(input)) {
    return false;
  }
  return true;
}

// @region[claude-agent-sdk-typescript-adapter]
// @region[claude-agent-sdk-agent-setup]
export async function runWithClaudeAgentSdk({
  input,
  emit,
  runId,
  threadId,
  systemPrompt,
  toolSchemas,
  initialState,
  model,
  executeTool,
  forwardedHeaders,
}: {
  input: RunAgentInput;
  emit: Emit;
  runId: string;
  threadId: string;
  systemPrompt: string;
  toolSchemas: Anthropic.Tool[];
  initialState: Record<string, unknown>;
  model: string;
  executeTool: ExecuteTool;
  forwardedHeaders?: Record<string, string>;
}): Promise<void> {
  let state = { ...initialState };
  const pendingStateSnapshots: Record<string, unknown>[] = [];
  const backendToolServer = buildBackendToolServer({
    toolSchemas,
    emit,
    getState: () => state,
    setState: (nextState) => {
      state = nextState;
      pendingStateSnapshots.push(state);
    },
    executeTool,
  });

  const adapter = new ClaudeAgentAdapter({
    agentId: "claude-sdk-typescript",
    model: normalizeClaudeAgentSdkModel(model),
    systemPrompt,
    tools: [],
    mcpServers: backendToolServer.mcpServers,
    allowedTools: backendToolServer.allowedTools,
    permissionMode: "dontAsk",
    maxTurns: 10,
  });

  if (forwardedHeaders && Object.keys(forwardedHeaders).length > 0) {
    adapter.headers = forwardedHeaders;
  }

  const runInput: RunAgentInput = {
    ...input,
    runId,
    threadId,
    state: input.state ?? initialState,
  };

  await new Promise<void>((resolve) => {
    adapter.run(runInput).subscribe({
      next: (event) => {
        if (event.type === EventType.TOOL_CALL_RESULT) {
          const snapshot = pendingStateSnapshots.shift();
          if (snapshot) {
            emit({ type: EventType.STATE_SNAPSHOT, snapshot });
          }
        }
        emit(event);
      },
      error: (error) => {
        const message =
          error instanceof Error ? error.stack || error.message : String(error);
        emit({ type: EventType.RUN_ERROR, runId, threadId, message });
        resolve();
      },
      complete: () => resolve(),
    });
  });
}
// @endregion[claude-agent-sdk-agent-setup]

function buildBackendToolServer({
  toolSchemas,
  emit,
  getState,
  setState,
  executeTool,
}: {
  toolSchemas: Anthropic.Tool[];
  emit: Emit;
  getState: () => Record<string, unknown>;
  setState: (state: Record<string, unknown>) => void;
  executeTool: ExecuteTool;
}): {
  mcpServers?: Record<string, McpServerConfig>;
  allowedTools: string[];
} {
  if (toolSchemas.length === 0) {
    return { allowedTools: [] };
  }

  const tools = toolSchemas.map((schema) =>
    sdkTool(
      schema.name,
      schema.description ?? "",
      zodShapeFromJsonSchema(schema.input_schema),
      async (args) => {
        try {
          const result = await executeTool(
            schema.name,
            args as Record<string, unknown>,
            getState(),
            emit,
          );
          if (result.state) {
            setState(result.state);
          }
          return {
            content: [{ type: "text" as const, text: result.resultText }],
          };
        } catch (error) {
          const message =
            error instanceof Error ? error.message : String(error);
          return {
            content: [{ type: "text" as const, text: message }],
            isError: true,
          };
        }
      },
    ),
  );

  return {
    mcpServers: {
      [COPILOTKIT_MCP_SERVER_NAME]: createSdkMcpServer({
        name: COPILOTKIT_MCP_SERVER_NAME,
        version: "1.0.0",
        tools,
      }),
    },
    allowedTools: toolSchemas.map(
      (schema) => `${COPILOTKIT_TOOL_PREFIX}${schema.name}`,
    ),
  };
}

function zodShapeFromJsonSchema(
  schema: Anthropic.Tool.InputSchema,
): AnyZodRawShape {
  const properties =
    typeof schema === "object" && schema && "properties" in schema
      ? ((schema as { properties?: Record<string, unknown> }).properties ?? {})
      : {};
  const required = new Set(
    Array.isArray((schema as { required?: unknown }).required)
      ? ((schema as { required: string[] }).required ?? [])
      : [],
  );
  const shape: Record<string, ZodTypeAny> = {};
  for (const [name, propertySchema] of Object.entries(properties)) {
    const field = zodFromJsonSchema(propertySchema);
    shape[name] = required.has(name) ? field : field.optional();
  }
  return shape as unknown as AnyZodRawShape;
}

function zodFromJsonSchema(schema: unknown): ZodTypeAny {
  if (!schema || typeof schema !== "object") {
    return z.any();
  }
  const typed = schema as {
    type?: string;
    enum?: unknown[];
    items?: unknown;
    properties?: Record<string, unknown>;
    required?: string[];
  };
  if (
    typed.enum?.length &&
    typed.enum.every((value) => typeof value === "string")
  ) {
    const values = typed.enum as [string, ...string[]];
    return z.enum(values);
  }
  if (typed.type === "string") return z.string();
  if (typed.type === "number" || typed.type === "integer") return z.number();
  if (typed.type === "boolean") return z.boolean();
  if (typed.type === "array") return z.array(zodFromJsonSchema(typed.items));
  if (typed.type === "object" && typed.properties) {
    const required = new Set(typed.required ?? []);
    const shape: Record<string, ZodTypeAny> = {};
    for (const [name, propertySchema] of Object.entries(typed.properties)) {
      const field = zodFromJsonSchema(propertySchema);
      shape[name] = required.has(name) ? field : field.optional();
    }
    return z.object(shape);
  }
  return z.any();
}

function hasStructuredUserContent(input: RunAgentInput): boolean {
  for (const message of input.messages ?? []) {
    const content = (message as { content?: unknown }).content;
    if (message.role !== "user" || !Array.isArray(content)) continue;
    for (const part of content) {
      if (!part || typeof part !== "object") continue;
      const type = (part as { type?: unknown }).type;
      if (type && type !== "text") return true;
    }
  }
  return false;
}

function hasHeader(headers: Record<string, string>, name: string): boolean {
  return Object.keys(headers).some((key) => key.toLowerCase() === name);
}

function normalizeClaudeAgentSdkModel(model: string): string {
  return model === "claude-sonnet-4.6" ? "claude-sonnet-4-6" : model;
}
// @endregion[claude-agent-sdk-typescript-adapter]
